File: llm/src/llm/Privatellm.py
from dotenv import load_dotenv
import os
from typing import Any

load_dotenv()

class PrivateLLM:
    def __init__(self) -> None:
        self.model = str( os.getenv("DEPLOYMENT_NAME") )
        self.api_key = str( os.getenv("OPENAI_API_KEY") )
        self.api_version = str( os.getenv("OPENAI_API_VERSION") )
        self.resource = str( os.getenv("RESOURCE") )

    def PromptForBulletPoint(self, prompt: str, type: str) -> Any:
        try:
            import requests
            import json

            chat = []
            chat.append(self.create_system_message(type))

            message_user = self.create_user_message(prompt)

            print(message_user)

            chat.append(message_user)

            url = 'https://{0}/openai/deployments/{1}/chat/completions?api-version={2}'.format(self.resource,
                                                                                               self.model, self.api_version)
            headers = {
                'Content-Type': 'application/json',
                'api-key': self.api_key
            }
            body_data = {
                "messages": chat
            }

            print(body_data)

            response = requests.post(url, headers=headers, json=body_data)
            text = response.text
            parsed_json = json.loads(text)

            history = parsed_json['choices'][0]['message']
            chat.append(history)

            return {"history": chat[-1]}

        except Exception as error:
            raise Exception(error)


    def create_user_message(self, message):
        return {"role": "user", "content": "{0}".format(message)}

    def create_assistant_message(self, message):
        return {"role": "assistant", "content": "{0}".format(message)}

    def create_system_message(self, type: str):
        return {"role": "assistant", "content": f"Você é uma IA especialidada em resumir os prompts que o usuário te passar e retornar em formato markdow, mas formatando os pontos principais de acordo com o que inferir no tipo: {type}"}

File: src/server/external/shared/PrivateLLM.go
package shared

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/gofiber/contrib/websocket"
	"github.com/joho/godotenv"
)

func init() {
	godotenv.Load()
}

type PrivateLLM struct {
	Model      string
	APIKey     string
	APIVersion string
	Resource   string
}

func NewPrivateLLM() *PrivateLLM {
	return &PrivateLLM{
		Model:      os.Getenv("DEPLOYMENT_NAME"),
		APIKey:     os.Getenv("OPENAI_API_KEY"),
		APIVersion: os.Getenv("OPENAI_API_VERSION"),
		Resource:   os.Getenv("RESOURCE"),
	}
}

// Message represents a chat message.
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// Response represents the JSON response from the OpenAI API.
type Response struct {
	Choices []struct {
		Message Message `json:"message"`
	} `json:"choices"`
}

// PromptForBulletPoint generates a response for the given prompt and type.
func (llm *PrivateLLM) PromptForBulletPoint(prompt string, typeStr string, c *websocket.Conn) (string, error) {
	chat := []Message{
		llm.createSystemMessage(typeStr),
		llm.createUserMessage(prompt),
	}

	url := fmt.Sprintf("https://%s/openai/deployments/%s/chat/completions?api-version=%s", llm.Resource, llm.Model, llm.APIVersion)
	headers := map[string]string{
		"Content-Type": "application/json",
		"api-key":      llm.APIKey,
	}
	bodyData := map[string][]Message{
		"messages": chat,
	}

	jsonBody, err := json.Marshal(bodyData)
	if err != nil {
		return "", err
	}

	response, err := postJSON(url, headers, jsonBody)
	if err != nil {
		return "", err
	}

	var parsedJSON Response
	if err := json.Unmarshal(response, &parsedJSON); err != nil {
		return "", err
	}

	history := parsedJSON.Choices[0].Message.Content
	chat = append(chat, Message{Role: "assistant", Content: history})

	fmt.Println(chat[len(chat)-1].Content)

	type Response struct {
		Response string `json:"response"`
	}

	resp := Response{
		Response: chat[len(chat)-1].Content,
	}

	if err = c.WriteJSON(resp); err != nil {
		log.Println(err)
	}

	return chat[len(chat)-1].Content, nil
}

func (llm *PrivateLLM) createUserMessage(message string) Message {
	return Message{Role: "user", Content: message}
}

func (llm *PrivateLLM) createSystemMessage(typeStr string) Message {
	return Message{Role: "assistant", Content: fmt.Sprintf("Você é uma IA especializada em resumir os prompts que o usuário te passar e retornar em formato markdown, mas formatando os pontos principais de acordo com o que inferir no tipo: %s", typeStr)}
}

func postJSON(url string, headers map[string]string, body []byte) ([]byte, error) {
	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}

	// Set headers
	for key, value := range headers {
		req.Header.Set(key, value)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP request failed with status code %d", resp.StatusCode)
	}

	responseBody := new(bytes.Buffer)
	_, err = responseBody.ReadFrom(resp.Body)
	if err != nil {
		return nil, err
	}

	return responseBody.Bytes(), nil
}
